File: packages/wasm-api/src/terminal.ts
/**
 * TypeScript wrapper for the Ghostty Terminal WASM API
 */

import type {
  GhosttyWasmExports,
  TerminalCell,
  CursorPosition,
  TerminalSize,
  ScrollbackInfo,
} from "./types.js";
import { WasmMemoryManager } from "./memory.js";

/**
 * High-level wrapper for the Ghostty Terminal WASM API
 */
export class GhosttyTerminal {
  private terminalPtr: number;
  private memory: WasmMemoryManager;
  private _cols: number;
  private _rows: number;

  constructor(
    private exports: GhosttyWasmExports,
    cols: number,
    rows: number
  ) {
    this.memory = new WasmMemoryManager(exports);
    this._cols = cols;
    this._rows = rows;

    // Create terminal instance
    const termPtrPtr = this.memory.allocOpaque();
    const result = this.exports.ghostty_terminal_new(0, cols, rows, termPtrPtr);

    if (result !== 0) {
      this.memory.freeOpaque(termPtrPtr);
      throw new Error(`Failed to create terminal: error code ${result}`);
    }

    this.terminalPtr = this.memory.readU32(termPtrPtr);
    this.memory.freeOpaque(termPtrPtr);
  }

  /**
   * Get the terminal pointer (for advanced use cases)
   */
  getPtr(): number {
    return this.terminalPtr;
  }

  /**
   * Get the memory manager (for advanced use cases)
   */
  getMemory(): WasmMemoryManager {
    return this.memory;
  }

  /**
   * Get the current terminal size
   */
  getSize(): TerminalSize {
    return {
      cols: this._cols,
      rows: this._rows,
    };
  }

  /**
   * Write data to the terminal
   */
  write(data: string | Uint8Array): void {
    const bytes = typeof data === "string" ? new TextEncoder().encode(data) : data;

    const dataPtr = this.memory.allocU8Array(bytes.length);
    this.memory.copyToMemory(dataPtr, bytes);

    const result = this.exports.ghostty_terminal_write(
      this.terminalPtr,
      dataPtr,
      bytes.length
    );

    this.memory.freeU8Array(dataPtr);

    if (result !== 0) {
      throw new Error(`Terminal write failed: error code ${result}`);
    }
  }

  /**
   * Resize the terminal
   */
  resize(cols: number, rows: number): void {
    const result = this.exports.ghostty_terminal_resize(
      this.terminalPtr,
      cols,
      rows
    );

    if (result !== 0) {
      throw new Error(`Terminal resize failed: error code ${result}`);
    }

    this._cols = cols;
    this._rows = rows;
  }

  /**
   * Clear the terminal
   */
  clear(): void {
    this.exports.ghostty_terminal_clear(this.terminalPtr);
  }

  /**
   * Reset the terminal
   */
  reset(): void {
    this.exports.ghostty_terminal_reset(this.terminalPtr);
  }

  /**
   * Get cursor position
   */
  getCursor(): CursorPosition {
    const xPtr = this.memory.allocU16();
    const yPtr = this.memory.allocU16();

    this.exports.ghostty_terminal_get_cursor(this.terminalPtr, xPtr, yPtr);

    const x = this.memory.readU16(xPtr);
    const y = this.memory.readU16(yPtr);

    this.memory.freeU16(xPtr);
    this.memory.freeU16(yPtr);

    return { x, y };
  }

  /**
   * Get cell data at viewport coordinates
   */
  getCellViewport(x: number, y: number): TerminalCell | null {
    const codepointPtr = this.memory.allocU32();
    const fgRPtr = this.memory.allocU8();
    const fgGPtr = this.memory.allocU8();
    const fgBPtr = this.memory.allocU8();
    const bgRPtr = this.memory.allocU8();
    const bgGPtr = this.memory.allocU8();
    const bgBPtr = this.memory.allocU8();
    const boldPtr = this.memory.allocU8();
    const italicPtr = this.memory.allocU8();
    const underlinePtr = this.memory.allocU8();

    const hasCell = this.exports.ghostty_terminal_get_cell_viewport(
      this.terminalPtr,
      x,
      y,
      codepointPtr,
      fgRPtr,
      fgGPtr,
      fgBPtr,
      bgRPtr,
      bgGPtr,
      bgBPtr,
      boldPtr,
      italicPtr,
      underlinePtr
    );

    if (!hasCell) {
      this.memory.freeU32(codepointPtr);
      this.memory.freeU8(fgRPtr);
      this.memory.freeU8(fgGPtr);
      this.memory.freeU8(fgBPtr);
      this.memory.freeU8(bgRPtr);
      this.memory.freeU8(bgGPtr);
      this.memory.freeU8(bgBPtr);
      this.memory.freeU8(boldPtr);
      this.memory.freeU8(italicPtr);
      this.memory.freeU8(underlinePtr);
      return null;
    }

    const cell: TerminalCell = {
      codepoint: this.memory.readU32(codepointPtr),
      fg: {
        r: this.memory.readU8(fgRPtr),
        g: this.memory.readU8(fgGPtr),
        b: this.memory.readU8(fgBPtr),
      },
      bg: {
        r: this.memory.readU8(bgRPtr),
        g: this.memory.readU8(bgGPtr),
        b: this.memory.readU8(bgBPtr),
      },
      bold: this.memory.readU8(boldPtr) !== 0,
      italic: this.memory.readU8(italicPtr) !== 0,
      underline: this.memory.readU8(underlinePtr) !== 0,
    };

    this.memory.freeU32(codepointPtr);
    this.memory.freeU8(fgRPtr);
    this.memory.freeU8(fgGPtr);
    this.memory.freeU8(fgBPtr);
    this.memory.freeU8(bgRPtr);
    this.memory.freeU8(bgGPtr);
    this.memory.freeU8(bgBPtr);
    this.memory.freeU8(boldPtr);
    this.memory.freeU8(italicPtr);
    this.memory.freeU8(underlinePtr);

    return cell;
  }

  /**
   * Get all cells in the viewport at once.
   * Returns an array of TerminalCell objects in row-major order.
   */
  getAllCellsViewport(): TerminalCell[] {
    const size = this.getSize();
    const totalCells = size.cols * size.rows;
    const cellSize = 14; // bytes per cell
    const bufferSize = totalCells * cellSize;

    // Allocate buffer in WASM memory
    const bufferPtr = this.memory.allocU8Array(bufferSize);

    const success = this.exports.ghostty_terminal_get_all_cells_viewport(
      this.terminalPtr,
      bufferPtr
    );

    if (!success) {
      this.memory.freeU8Array(bufferPtr);
      return [];
    }

    // Read cell data from buffer
    const buffer = this.memory.copyFromMemory(bufferPtr, bufferSize);
    const dataView = new DataView(buffer.buffer, buffer.byteOffset, bufferSize);
    const cells: TerminalCell[] = [];

    for (let i = 0; i < totalCells; i++) {
      const offset = i * cellSize;
      const codepoint = dataView.getUint32(offset, true); // true = little endian
      const fgR = buffer[offset + 4];
      const fgG = buffer[offset + 5];
      const fgB = buffer[offset + 6];
      const bgR = buffer[offset + 7];
      const bgG = buffer[offset + 8];
      const bgB = buffer[offset + 9];
      const bold = buffer[offset + 10] !== 0;
      const italic = buffer[offset + 11] !== 0;
      const underline = buffer[offset + 12] !== 0;

      cells.push({
        codepoint,
        fg: { r: fgR, g: fgG, b: fgB },
        bg: { r: bgR, g: bgG, b: bgB },
        bold,
        italic,
        underline,
      });
    }

    this.memory.freeU8Array(bufferPtr);
    return cells;
  }

  /**
   * Get terminal title
   */
  getTitle(): string {
    const titlePtrPtr = this.memory.allocOpaque();
    const titleLenPtr = this.memory.allocUsize();

    this.exports.ghostty_terminal_get_title(
      this.terminalPtr,
      titlePtrPtr,
      titleLenPtr
    );

    const titlePtr = this.memory.readU32(titlePtrPtr);
    const titleLen = this.memory.readUsize(titleLenPtr);

    this.memory.freeOpaque(titlePtrPtr);
    this.memory.freeUsize(titleLenPtr);

    if (titleLen === 0) {
      return "";
    }

    return this.memory.readString(titlePtr, titleLen);
  }

  /**
   * Get scrollback information
   */
  getScrollback(): ScrollbackInfo {
    const totalRowsPtr = this.memory.allocUsize();
    const viewportOffsetPtr = this.memory.allocUsize();
    const visibleRowsPtr = this.memory.allocUsize();

    this.exports.ghostty_terminal_get_scrollback(
      this.terminalPtr,
      totalRowsPtr,
      viewportOffsetPtr,
      visibleRowsPtr
    );

    const info: ScrollbackInfo = {
      totalRows: this.memory.readUsize(totalRowsPtr),
      viewportOffset: this.memory.readUsize(viewportOffsetPtr),
      visibleRows: this.memory.readUsize(visibleRowsPtr),
    };

    this.memory.freeUsize(totalRowsPtr);
    this.memory.freeUsize(viewportOffsetPtr);
    this.memory.freeUsize(visibleRowsPtr);

    return info;
  }

  /**
   * Set viewport offset for scrollback
   */
  setViewportOffset(offset: number): void {
    this.exports.ghostty_terminal_set_viewport_offset(this.terminalPtr, offset);
  }

  /**
   * Free the terminal
   */
  destroy(): void {
    if (this.terminalPtr) {
      this.exports.ghostty_terminal_free(this.terminalPtr);
      this.terminalPtr = 0;
    }
  }
}

File: src/terminal/c/terminal.zig
const std = @import("std");
const assert = std.debug.assert;
const builtin = @import("builtin");
const lib_alloc = @import("../../lib/allocator.zig");
const CAllocator = lib_alloc.Allocator;
const Terminal = @import("../Terminal.zig");
const stream = @import("../stream.zig");
const Result = @import("result.zig").Result;
const Page = @import("../page.zig").Page;
const Cell = @import("../page.zig").Cell;
const color = @import("../color.zig");

const log = std.log.scoped(.terminal_c);

/// Custom handler for capturing title changes
const TitleTrackingHandler = struct {
    terminal: *Terminal,
    title: []u8,
    title_len: usize,
    alloc: std.mem.Allocator,

    fn init(alloc: std.mem.Allocator, terminal: *Terminal) !TitleTrackingHandler {
        const title_buf = try alloc.alloc(u8, 256); // Initial buffer size
        return .{
            .terminal = terminal,
            .title = title_buf,
            .title_len = 0,
            .alloc = alloc,
        };
    }

    pub fn deinit(self: *TitleTrackingHandler) void {
        self.alloc.free(self.title);
    }

    pub fn vt(
        self: *TitleTrackingHandler,
        comptime action: stream.Action.Tag,
        value: stream.Action.Value(action),
    ) !void {
        // Forward all events to the terminal's handler
        var handler = self.terminal.vtHandler();
        try handler.vt(action, value);

        // Also capture title changes
        if (action == .window_title) {
            const title = value.title;
            if (title.len > self.title.len) {
                self.alloc.free(self.title);
                self.title = try self.alloc.alloc(u8, title.len);
            }
            @memcpy(self.title[0..title.len], title);
            self.title_len = title.len;
        }
    }
};

const TitleTrackingStream = stream.Stream(TitleTrackingHandler);

/// Wrapper around Terminal that tracks allocator for C API usage.
const TerminalWrapper = struct {
    terminal: Terminal,
    stream: TitleTrackingStream,
    alloc: std.mem.Allocator,
};

/// C: GhosttyTerminal
pub const CTerminal = ?*TerminalWrapper;

/// Create a new terminal instance.
pub fn new(
    alloc_: ?*const CAllocator,
    cols: u16,
    rows: u16,
    result: *CTerminal,
) callconv(.c) Result {
    const alloc = lib_alloc.default(alloc_);
    const ptr = alloc.create(TerminalWrapper) catch
        return .out_of_memory;

    ptr.terminal = Terminal.init(alloc, .{
        .cols = cols,
        .rows = rows,
    }) catch {
        alloc.destroy(ptr);
        return .out_of_memory;
    };

    const handler = TitleTrackingHandler.init(alloc, &ptr.terminal) catch {
        ptr.terminal.deinit(alloc);
        alloc.destroy(ptr);
        return .out_of_memory;
    };

    ptr.stream = TitleTrackingStream.initAlloc(alloc, handler);
    ptr.alloc = alloc;
    result.* = ptr;
    return .success;
}

/// Free a terminal instance.
pub fn free(terminal_: CTerminal) callconv(.c) void {
    const wrapper = terminal_ orelse return;
    const alloc = wrapper.alloc;
    wrapper.stream.deinit();
    wrapper.terminal.deinit(alloc);
    alloc.destroy(wrapper);
}

/// Feed data to the terminal parser.
pub fn write(
    terminal_: CTerminal,
    data: [*]const u8,
    len: usize,
) callconv(.c) Result {
    const wrapper = terminal_ orelse return .invalid_value;
    const slice = data[0..len];
    wrapper.stream.nextSlice(slice) catch return .out_of_memory;
    return .success;
}

/// Get the terminal screen dimensions.
pub fn getSize(
    terminal_: CTerminal,
    cols: *u16,
    rows: *u16,
) callconv(.c) void {
    const wrapper = terminal_ orelse return;
    cols.* = wrapper.terminal.cols;
    rows.* = wrapper.terminal.rows;
}

/// Get cursor position.
pub fn getCursor(
    terminal_: CTerminal,
    x: *u16,
    y: *u16,
) callconv(.c) void {
    const wrapper = terminal_ orelse return;
    const screen = wrapper.terminal.screen;
    x.* = @intCast(screen.cursor.x);
    y.* = @intCast(screen.cursor.y);
}

/// Get a cell at a specific position in active area coordinates.
/// x and y are relative to the active area (0,0 is top-left of active area).
/// The active area is where the cursor lives and where programs can write.
/// Returns false if position is out of bounds.
pub fn getCell(
    terminal_: CTerminal,
    x: u16,
    y: u16,
    codepoint: *u32,
    fg_r: *u8,
    fg_g: *u8,
    fg_b: *u8,
    bg_r: *u8,
    bg_g: *u8,
    bg_b: *u8,
    bold: *bool,
    italic: *bool,
    underline: *bool,
) callconv(.c) bool {
    const wrapper = terminal_ orelse return false;
    const screen = &wrapper.terminal.screen;

    if (y >= wrapper.terminal.rows or x >= wrapper.terminal.cols) {
        return false;
    }

    const pin = screen.pages.pin(.{ .active = .{ .x = x, .y = y } }) orelse return false;

    const rac = pin.rowAndCell();
    const cell = rac.cell;

    // Get codepoint based on content tag
    codepoint.* = switch (cell.content_tag) {
        .codepoint, .codepoint_grapheme => cell.content.codepoint,
        else => 0,
    };

    // Get style from pin
    const cell_style = pin.style(cell);

    // Get color palette
    const palette = &wrapper.terminal.colors.palette.current;

    // Get default colors (fallback to white on black if not set)
    const default_fg = wrapper.terminal.colors.foreground.get() orelse color.RGB{ .r = 255, .g = 255, .b = 255 };
    const default_bg = wrapper.terminal.colors.background.get() orelse color.RGB{ .r = 0, .g = 0, .b = 0 };

    // Get foreground color
    const fg_color = cell_style.fg(.{
        .default = default_fg,
        .palette = palette,
        .bold = null,
    });
    fg_r.* = fg_color.r;
    fg_g.* = fg_color.g;
    fg_b.* = fg_color.b;

    // Get background color
    const bg_color = cell_style.bg(cell, palette) orelse default_bg;
    bg_r.* = bg_color.r;
    bg_g.* = bg_color.g;
    bg_b.* = bg_color.b;

    // Get styles
    bold.* = cell_style.flags.bold;
    italic.* = cell_style.flags.italic;
    underline.* = cell_style.flags.underline != .none;

    return true;
}

/// Get a cell at a specific position in viewport coordinates.
/// x and y are relative to the current viewport (0,0 is top-left of visible area).
/// This respects the current scroll position.
/// Returns false if position is out of bounds.
pub fn getCellViewport(
    terminal_: CTerminal,
    x: u16,
    y: u16,
    codepoint: *u32,
    fg_r: *u8,
    fg_g: *u8,
    fg_b: *u8,
    bg_r: *u8,
    bg_g: *u8,
    bg_b: *u8,
    bold: *bool,
    italic: *bool,
    underline: *bool,
) callconv(.c) bool {
    const wrapper = terminal_ orelse return false;
    const screen = &wrapper.terminal.screen;

    if (y >= wrapper.terminal.rows or x >= wrapper.terminal.cols) {
        return false;
    }

    const pin = screen.pages.pin(.{ .viewport = .{ .x = x, .y = y } }) orelse return false;

    const rac = pin.rowAndCell();
    const cell = rac.cell;

    // Get codepoint based on content tag
    codepoint.* = switch (cell.content_tag) {
        .codepoint, .codepoint_grapheme => cell.content.codepoint,
        else => 0,
    };

    // Get style from pin
    const cell_style = pin.style(cell);

    // Get color palette
    const palette = &wrapper.terminal.colors.palette.current;

    // Get default colors (fallback to white on black if not set)
    const default_fg = wrapper.terminal.colors.foreground.get() orelse color.RGB{ .r = 255, .g = 255, .b = 255 };
    const default_bg = wrapper.terminal.colors.background.get() orelse color.RGB{ .r = 0, .g = 0, .b = 0 };

    // Get foreground color
    const fg_color = cell_style.fg(.{
        .default = default_fg,
        .palette = palette,
        .bold = null,
    });
    fg_r.* = fg_color.r;
    fg_g.* = fg_color.g;
    fg_b.* = fg_color.b;

    // Get background color
    const bg_color = cell_style.bg(cell, palette) orelse default_bg;
    bg_r.* = bg_color.r;
    bg_g.* = bg_color.g;
    bg_b.* = bg_color.b;

    // Get styles
    bold.* = cell_style.flags.bold;
    italic.* = cell_style.flags.italic;
    underline.* = cell_style.flags.underline != .none;

    return true;
}

/// Clear the terminal screen.
pub fn clear(terminal_: CTerminal) callconv(.c) void {
    const wrapper = terminal_ orelse return;
    // Send clear screen sequence
    wrapper.stream.nextSlice("\x1b[2J") catch {};
    wrapper.stream.nextSlice("\x1b[H") catch {};
}

/// Reset the terminal to initial state.
pub fn reset(terminal_: CTerminal) callconv(.c) void {
    const wrapper = terminal_ orelse return;
    // Send RIS (Reset to Initial State) sequence
    wrapper.stream.nextSlice("\x1bc") catch {};
}

/// Resize the terminal to new dimensions.
pub fn resize(
    terminal_: CTerminal,
    cols: u16,
    rows: u16,
) callconv(.c) Result {
    const wrapper = terminal_ orelse return .invalid_value;
    wrapper.terminal.resize(wrapper.alloc, cols, rows) catch return .out_of_memory;
    return .success;
}

/// Get the current window title.
/// Returns a pointer to the title string and its length.
/// The pointer is valid until the next call to write() or free().
pub fn getTitle(
    terminal_: CTerminal,
    title_ptr: *[*]const u8,
    title_len: *usize,
) callconv(.c) void {
    const wrapper = terminal_ orelse {
        title_ptr.* = "";
        title_len.* = 0;
        return;
    };

    title_ptr.* = wrapper.stream.handler.title.ptr;
    title_len.* = wrapper.stream.handler.title_len;
}

/// Get scrollback information.
/// Returns the total number of rows (including scrollback and visible rows),
/// the current viewport offset, and the number of visible rows.
pub fn getScrollback(
    terminal_: CTerminal,
    total_rows: *usize,
    viewport_offset: *usize,
    visible_rows: *usize,
) callconv(.c) void {
    const wrapper = terminal_ orelse {
        total_rows.* = 0;
        viewport_offset.* = 0;
        visible_rows.* = 0;
        return;
    };

    const screen = &wrapper.terminal.screen;
    const sb = screen.pages.scrollbar();
    total_rows.* = sb.total;
    viewport_offset.* = sb.offset;
    visible_rows.* = sb.len;
}

/// Set the viewport offset for scrolling.
/// offset is the number of rows from the top of the scrollback buffer.
/// Returns success if the viewport was updated.
pub fn setViewportOffset(
    terminal_: CTerminal,
    offset: usize,
) callconv(.c) Result {
    const wrapper = terminal_ orelse return .invalid_value;
    const screen = &wrapper.terminal.screen;

    // Use the PageList scroll function to set viewport to the row offset
    screen.pages.scroll(.{ .row = offset });

    return .success;
}

/// Get all cells in the viewport at once.
/// Writes cell data to the provided buffer in row-major order (left-to-right, top-to-bottom).
/// Each cell is 14 bytes: codepoint (u32), fg_r (u8), fg_g (u8), fg_b (u8),
/// bg_r (u8), bg_g (u8), bg_b (u8), bold (u8), italic (u8), underline (u8), padding (u8).
/// buffer must be at least cols * rows * 14 bytes.
/// Returns false if buffer is null or terminal is invalid.
pub fn getAllCellsViewport(
    terminal_: CTerminal,
    buffer: [*]u8,
) callconv(.c) bool {
    const wrapper = terminal_ orelse return false;
    const screen = &wrapper.terminal.screen;
    const cols = wrapper.terminal.cols;
    const rows = wrapper.terminal.rows;

    // Get color palette
    const palette = &wrapper.terminal.colors.palette.current;

    // Get default colors (fallback to white on black if not set)
    const default_fg = wrapper.terminal.colors.foreground.get() orelse color.RGB{ .r = 255, .g = 255, .b = 255 };
    const default_bg = wrapper.terminal.colors.background.get() orelse color.RGB{ .r = 0, .g = 0, .b = 0 };

    var offset: usize = 0;
    const cell_size = 14;

    for (0..rows) |y| {
        for (0..cols) |x| {
            const pin = screen.pages.pin(.{ .viewport = .{ .x = @intCast(x), .y = @intCast(y) } }) orelse {
                // Out of bounds cell - write zeros
                @memset(buffer[offset..offset + cell_size], 0);
                offset += cell_size;
                continue;
            };

            const rac = pin.rowAndCell();
            const cell = rac.cell;

            // Get codepoint based on content tag
            const codepoint_val: u32 = switch (cell.content_tag) {
                .codepoint, .codepoint_grapheme => cell.content.codepoint,
                else => 0,
            };

            // Get style from pin
            const cell_style = pin.style(cell);

            // Get foreground color
            const fg_color = cell_style.fg(.{
                .default = default_fg,
                .palette = palette,
                .bold = null,
            });

            // Get background color
            const bg_color = cell_style.bg(cell, palette) orelse default_bg;

            // Write cell data to buffer
            const cell_ptr = @as(*[14]u8, @ptrCast(buffer + offset));
            // Write codepoint as u32 (little-endian)
            std.mem.writeInt(u32, cell_ptr[0..4], codepoint_val, .little);
            cell_ptr[4] = fg_color.r;
            cell_ptr[5] = fg_color.g;
            cell_ptr[6] = fg_color.b;
            cell_ptr[7] = bg_color.r;
            cell_ptr[8] = bg_color.g;
            cell_ptr[9] = bg_color.b;
            cell_ptr[10] = if (cell_style.flags.bold) 1 else 0;
            cell_ptr[11] = if (cell_style.flags.italic) 1 else 0;
            cell_ptr[12] = if (cell_style.flags.underline != .none) 1 else 0;
            cell_ptr[13] = 0; // padding

            offset += cell_size;
        }
    }

    return true;
}
